### PR TITLE
Simpler screencopy

### PR DESCRIPTION
**Pull request by any1**
This simplifies screencopy by copying pixels to shm in the commit callback rather than the pre-commit callback, so both dmabuf and shm buffers are handled in commit.

Because commit always happens after pre-commit, damage accumulation is also simplified because we no longer need to ascertain whether the damage callback or the frame-copy callback happened first.

Supersedes #2591

### CODE REVIEW

**Comment by emersion** [comment]
I haven't looked in detail yet, but

>renderer: Add buffer argument to read_pixels

I think we can just `bind_buffer`, `read_pixels`, then `bind_buffer(NULL)` to avoid this change?

**Comment by any1** [comment]
> I haven't looked in detail yet, but
> 
> > renderer: Add buffer argument to read_pixels
> 
> I think we can just `bind_buffer`, `read_pixels`, then `bind_buffer(NULL)` to avoid this change?

Hmm, yeah, you're most likely right. I wasn't thinking in terms of the "global picture"; just what would be needed to call `read_pixels` from pretty much anywhere at any time. This change should make it possible to skip scheduling a frame and waiting for the commit event when the user doesn't want to wait for damage or when there is already damage. But maybe that's not really a useful "optimisation".

**Comment by any1** [comment]
I tried using `bind_buffer`, but it failed on `assert(wlr_egl_is_current(renderer->egl))`. If we want to use these functions, we need to (at least) change them so that they set an egl context and they should probably also save/restore the old egl context.

**Comment by emersion** [comment]
>we need to (at least) change them so that they set an egl context

I think this would be a good change, and it's also needed to remove the EGL dependency from the backends. Would you mind opening a PR for this?

> save/restore the old egl context

I think saving/restoring makes sense when we need to temporarily set the EGL context in a single function (like `blit_dmabuf`). However I'm a little worried about saving in one function call and restoring in another one. This could cause issues if the saved EGL objects become invalid, for instance.

**Comment by any1** [comment]
> > we need to (at least) change them so that they set an egl context
> 
> I think this would be a good change, and it's also needed to remove the EGL dependency from the backends. Would you mind opening a PR for this?

Yeah, I'll give it a go

**Comment by any1** [comment]
We are putting this on hold until #2505 is merged. Specifically https://github.com/swaywm/wlroots/pull/2505/commits/47f59dd3042891903f2682b9f3e3cc93afe07ef5.

**Comment by emersion** [comment]
#2505 has been merged, I think now would be a good time to revisit this. https://github.com/swaywm/wlroots/pull/2955 adds `wlr_output.front_buffer`.

**Comment by emersion** [comment]
Ah, still one missing piece is the DRM backend, which still uses the old infrastructure. This will be fixed in https://github.com/swaywm/wlroots/pull/2903.

**Comment by emersion** [comment]
@any1, would you be willing to continue work on this?

**Comment by any1** [comment]
> @any1, would you be willing to continue work on this?

Yeah, sure.

**Comment by emersion** [comment]
We should now be able to listen to output commits exclusively, check for `event->committed & WLR_OUTPUT_STATE_BUFFER`, then grab `output->front_buffer`.

**Comment by any1** [comment]
I believe I have adapted my previous changes to master now. It does seem to work with wayvnc on both the wayland and headless backends at least.

File: types/wlr_screencopy_v1.c
Comment (emersion) [inline types/wlr_screencopy_v1.c:22]: This can probably be removed?

File: types/wlr_screencopy_v1.c
Comment (emersion) [inline types/wlr_screencopy_v1.c:211]: We should check for errors here

File: types/wlr_screencopy_v1.c
Comment (emersion) [inline types/wlr_screencopy_v1.c:273]: I think we can assume that if an output is enabled, it'll always have a buffer.

File: types/wlr_screencopy_v1.c
Comment (emersion) [inline types/wlr_screencopy_v1.c:268]: I think this is already done by the caller.

File: types/wlr_screencopy_v1.c
Comment (any1) [reply types/wlr_screencopy_v1.c:22]: The precommit handler? It's what accumulates damage per output on the client object.

Comment (emersion) [reply types/wlr_screencopy_v1.c:22]: Ah, sorry, not this one. But `wlr_screencopy_frame_v1.output_precommit` can be removed, now that the `wl_signal_add` call has been removed from `frame_handle_copy`.

**Comment by emersion** [comment]
Ready to merge with that one last cleanup comment fixed

File: types/wlr_screencopy_v1.c
Comment (any1) [reply types/wlr_screencopy_v1.c:22]: `wlr_screencopy_frame_v1.output_precommit` is removed now

**Review by emersion — APPROVED** [review]
Thanks!

**Pull request merged**
